include/itkCLEDataManager.h

```diff
@@ -0,0 +1,160 @@
+{
+/** \class CLEDataManager
+ * \brief CLE memory manager implemented using OpenCL. Required by CLEImage class.
+ *
+ * This class serves as a base class for CLEsperanto data container for CLEImage class,
+ * which is similar to ImageBase class for Image class. However, all the image-related
+ * meta data will be already stored in image class (parent of CLEImage), therefore
+ * we did not name it CLEImageBase. Rather, this class is a CLE-specific data manager
+ * that provides functionalities for CPU-GPU data synchronization and grafting GPU data.
+ *
+ * \ingroup ITKCLE
+ */
+template <typename TPixel, unsigned int VDim>
+class CLEDataManager : public Object // DataObject//
+{
+  template<typename,unsigned int> friend class CLEImage;
+
+public:
+  ITK_DISALLOW_COPY_AND_MOVE(CLEDataManager);
+
+  using Self = CLEDataManager;
+  using SelfTemplate = CLEDataManager<TPixel, VDim>;
+  using Superclass = Object;
+  using Pointer = SmartPointer<Self>;
+  using ConstPointer = SmartPointer<const Self>;
+
+  itkNewMacro(Self);
+  itkTypeMacro(CLEDataManager, Object);
+
+  using MutexHolderType = std::lock_guard<std::mutex>;
+
+  void
+  SetImagePointer(CLEImage<TPixel,VDim>* img)
+  {
+    m_Image = img;
+  }
+
+  CLEImage<TPixel, VDim>*
+  GetImagePointer()
+  {
+    return this->m_Image.GetPointer();
+  }
+
+  virtual void
+  Allocate();
+
+  virtual void
+  Initialize();
+
+  //void
+  //SetCPUDirtyFlag(bool isDirty)
+  //{
+  //  m_IsCPUBufferDirty = isDirty;
+  //}
+  //
+  //void
+  //SetGPUDirtyFlag(bool isDirty)
+  //{
+  //  m_IsGPUBufferDirty = isDirty;
+  //}
+  //
+  ///** Make GPU up-to-date and mark CPU as dirty.
+  // * Call this function when you want to modify CPU data */
+  //void
+  //SetCPUBufferDirty();
+  //
+  ///** Make CPU up-to-date and mark GPU as dirty.
+  // * Call this function when you want to modify GPU data */
+  //void
+  //SetGPUBufferDirty()
+  //{
+  //  this->UpdateCPUBuffer();
+  //  m_IsGPUBufferDirty = true;
+  //}
+  //
+  //bool
+  //IsCPUBufferDirty() const
+  //{
+  //  return m_IsCPUBufferDirty;
+  //}
+  //
+  //bool
+  //IsGPUBufferDirty() const
+  //{
+  //  return m_IsGPUBufferDirty;
+  //}
+
+  /** Synchronize CPU and GPU buffers (using dirty flags) */
+  bool
+  Update();
+
+  /** actual GPU->CPU memory copy takes place here */
+  virtual void
+  UpdateCPUBuffer();
+
+  /** actual CPU->GPU memory copy takes place here */
+  virtual void
+  UpdateGPUBuffer();
+
+  /** Method for grafting the content of one CLEDataManager into another one */
+  virtual void
+  Graft(const CLEDataManager * data);
+
+protected:
+  CLEDataManager();
+  ~CLEDataManager() override;
+  void
+  PrintSelf(std::ostream & os, Indent indent) const override;
+
+private:
+  WeakPointer<CLEImage<TPixel, VDim>> m_Image; // WeakPointer has to be used here
+                                               // to avoid SmartPointer loop
+  std::array<size_t, 3> m_BufferShape;
+  cle::Object m_GPUBuffer;
+  cle::Clesperanto cle;
```

Comment:
`static`  (at least for now). CLContextManager that holds a static singleton (can create an in issue if needed for the next PR.